File: cpp/Array/0367_Valid_Perfect_Square.cpp
using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  /*
   * binary search
   * */
  bool isPerfectSquare(int num) {
    if (num == 1 || num == 4) return true;
    long long left = 0, right = num / 2;
    while (left < right) {
      long long mid = left + ((right - left) >> 1);
      if (num < mid * mid) right = mid;
      else if (num > mid * mid) left = mid + 1;
      else return true;
    }
    return false;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0367-Valid-Perfect-Square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num <= 1:
            return True
        left, right = 0, num // 2
        while left <= right:
            mid = (left + right) // 2
            mid_square = mid * mid
            if mid_square < num:
                left = mid + 1
            elif mid_square > num:
                right = mid - 1
            else:
                return True
        return False


s = Solution()
num = 16
print(s.isPerfectSquare(num))

num = 14
print(s.isPerfectSquare(num))
